src/app/application/services/data-source/DataSource.ts

```diff
@@ -0,0 +1,83 @@
+import CallableEventSource, { isClosed } from './CallableEventSource'
+import Router from './Router'
+import SharedPool from './SharedPool'
+
+export type DataSourceResponse<T> = {data: T | undefined, error: Error | undefined, refresh: () => void}
+export type Source = (params: Record<string, unknown>, source: {close: () => void}) => Promise<unknown>
+export type Sources = Record<string, Source>
+
+const get = (src: string, router: Router<Source>) => {
+  const [path, query] = src.split('?')
+  const queryParams = new URLSearchParams(query)
+  const route = router.match(path)
+  const _source = new CallableEventSource(async function * (this: CallableEventSource) {
+    while (true) {
+      this.readyState = 1
+      // latency
+      // await new Promise(resolve => setTimeout(resolve, 5000))
+      yield route.route({
+        ...{
+          offset: parseInt(queryParams.get('offset') || '0'),
+        },
+        ...route.params,
+      }, this)
+      if (!isClosed(this)) {
+        await new Promise(resolve => setTimeout(resolve, 5000))
+      }
+    }
+  })
+  return _source as EventSource
+}
+const destroy = (_src: string, source: EventSource) => {
+  if (source) {
+    source.close()
+  }
+}
+
+export class DataSource {
+  cache: Map<string, unknown> = new Map()
+  pool: SharedPool<string, EventSource>
+  constructor(requests: Sources) {
+    const requestRouter: Router<Source> = new Router(requests)
+
+    this.pool = new SharedPool<string, EventSource>(
+      (src: string) => {
+        return get(src, requestRouter)
+      },
+      (src: string, source: EventSource) => {
+        return destroy(src, source)
+      },
+    )
```

Comment:
The SharedPool manages request sharing and therefore response sharing and knows how to acquire (and initiate, if its the first instance) and teardown (if its the last instance) connections/Sources. SharedPool keeps a record of how many things are currently using a particular `Source`. Once nothing is using a `Source` anymore it gets torn down/destroyed.